compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4648,6 +4648,12 @@ private void setEventualTypeForExpression(BLangExpression expression,
         if (isSimpleWorkerReference(expression, data)) {
             return;
         }
+        if (expression.expectedType.tag != TypeTags.FUTURE) {
```

Comment:
I don't think this is the correct fix. 
1. We should give a better error if possible: indicating alternate wait action cannot be used as a wait-future-expr. 

Right now, the error for 

```ballerina
type WaitResult record {|
    boolean|error producer;
|};

function fn() {
    worker Producer1 returns boolean|error {
        return true;
    }

    worker Producer2 returns boolean|error {
        return true;
    }

    WaitResult _ = wait {producer: Producer1|Producer2};
}
```

is

```
ERROR [x.bal:(14:36,14:55)] incompatible types: expected 'future<(any|error)>', found eventual type '(future<(boolean|error)>|future<(boolean|error)>)' for wait future expression 'Producer1 | Producer2'
```

which doesn't quite make sense.

Here, `Producer1|Producer2` seems to actually be type-checked as an alternate wait action, which shouldn't be the case according to the current spec.

2. This will also disallow something like

```ballerina
type WaitResult record {|
    boolean|error producer;
|};

function fn() {
    worker Producer1 returns boolean|error {
        return true;
    }

    worker Producer2 returns boolean|error {
        return true;
    }

    future<boolean|error>|future<boolean|error> x = Producer1;
    WaitResult r2 = wait {producer: x};
}
```

Reply:
Created a spec issue for this. https://github.com/ballerina-platform/ballerina-spec/issues/1174

Reply:
Can we improve the error here? We basically need to say we expect an expression of a future type, right?

Right now we say

> incompatible types: expected 'future<(any|error)>', found eventual type '(future<(boolean|error)>|future<(boolean|error)>)' for wait future expression 'x'

which actually doesn't seem to be correct anyway. Eventual type with `future<T>` is `T`, but here we seem to be saying the eventual type is a future.

Reply:
Updated. Please check. 48185375e60cb5de6a89b97882502483cc1aea9c